### ISSUE TITLE
Deployment not completely scaling down old RS

### ISSUE THREAD

**Issue by bcorijn** [kubernetes/kubernetes#42571]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T07:30:54Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): `Debian GNU/Linux 8 (jessie)`
- **Kernel** (e.g. `uname -a`): `Linux ip-10-10-111-116 4.4.41-k8s #1 SMP Mon Jan 9 15:34:39 UTC 2017 x86_64 GNU/Linux`
- **Install tools**: Kops
- **Others**:

**What happened**:
I updated my deployment with a new image by using Helm and a custom chart (which has worked fine in the past) which manages 2 Deployments with both 2 replicas. Both Deployments created a new ReplicaSet, both came up online with 2 new pods, the old ReplicaSet was scaled down to 1 instance and then stayed there. My deployments now report as 3/2 and no further movement happens. This is the second time today this happens, I thought the first time might have been an odd error, but the next deploy has hit the same problem.

```
~ $ kubectl get deployments --namespace=prod
NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
ganalytics-prod          2         3         2            2           2d
ganalytics-prod-worker   2         3         2            2           2d
```
Deployment status hangs on terminating the last old pod:
```
~ $ kubectl rollout status deployment/ganalytics-prod --namespace=prod
Waiting for rollout to finish: 1 old replicas are pending termination...
```
Deployment description does not tell me much more eithe. It does show as one pod unavailable, but all pods are showing as running.
```
~ $ kubectl describe deployment ganalytics-prod --namespace=prod
Name:			ganalytics-prod
Namespace:		prod
CreationTimestamp:	Fri, 03 Mar 2017 15:18:49 +0100
Labels:			app=ganalytics
			chart=ganalytics-1
			heritage=Tiller
			release=ganalytics-prod
			role=main
Selector:		app=ganalytics,release=ganalytics-prod,role=main
Replicas:		2 updated | 2 total | 2 available | 1 unavailable
StrategyType:		RollingUpdate
MinReadySeconds:	25
RollingUpdateStrategy:	0 max unavailable, 100% max surge
Conditions:
  Type		Status	Reason
  ----		------	------
  Available 	True	MinimumReplicasAvailable
OldReplicaSets:	ganalytics-prod-2180515786 (1/1 replicas created)
NewReplicaSet:	ganalytics-prod-3154642894 (2/2 replicas created)
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  20m		20m		1	{deployment-controller }			Normal		ScalingReplicaSet	Scaled up replica set ganalytics-prod-3154642894 to 2
  20m		20m		1	{deployment-controller }			Normal		ScalingReplicaSet	Scaled down replica set ganalytics-prod-2180515786 to 1
```
```
~ $ kubectl get pods --namespace=prod
NAME                                      READY     STATUS    RESTARTS   AGE
ganalytics-prod-2180515786-tk1f5          1/1       Running   0          5h
ganalytics-prod-3154642894-2bn4n          1/1       Running   0          37m
ganalytics-prod-3154642894-rgxpp          1/1       Running   0          37m
```
Both replicaset report as all OK:
```
~ $ kubectl describe rs ganalytics-prod-2180515786 --namespace=prod
Name:		ganalytics-prod-2180515786
Namespace:	prod
Image(s):	woorank/ganalytics:v45
Selector:	app=ganalytics,pod-template-hash=2180515786,release=ganalytics-prod,role=main
Labels:		app=ganalytics
		chart=ganalytics-1
		heritage=Tiller
		pod-template-hash=2180515786
		release=ganalytics-prod
		role=main
		version=v45
Replicas:	1 current / 1 desired
Pods Status:	1 Running / 0 Waiting / 0 Succeeded / 0 Failed
No volumes.
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  36m		36m		1	{replicaset-controller }			Normal		SuccessfulDelete	Deleted pod: ganalytics-prod-2180515786-lbwjj
```
```
~ $ kubectl describe rs ganalytics-prod-3154642894 --namespace=prod
Name:		ganalytics-prod-3154642894
Namespace:	prod
Image(s):	woorank/ganalytics:v56
Selector:	app=ganalytics,pod-template-hash=3154642894,release=ganalytics-prod,role=main
Labels:		app=ganalytics
		chart=ganalytics-1
		heritage=Tiller
		pod-template-hash=3154642894
		release=ganalytics-prod
		role=main
		version=v56
Replicas:	2 current / 2 desired
Pods Status:	2 Running / 0 Waiting / 0 Succeeded / 0 Failed
No volumes.
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  21m		21m		1	{replicaset-controller }			Normal		SuccessfulCreate	Created pod: ganalytics-prod-3154642894-rgxpp
  21m		21m		1	{replicaset-controller }			Normal		SuccessfulCreate	Created pod: ganalytics-prod-3154642894-2bn4n

```
And lastly, kube-controller-manager  logs also show little special:
```
I0306 13:23:26.468174       6 replica_set.go:479] Too few "prod"/"ganalytics-prod-3154642894" replicas, need 2, creating 2
I0306 13:23:26.470697       6 event.go:217] Event(api.ObjectReference{Kind:"Deployment", Namespace:"prod", Name:"ganalytics-prod", UID:"524047a6-001c-11e7-bef2-0ab9b20c0a56", APIVersion:"extensions", ResourceVersion:"9142083", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set ganalytics-prod-3154642894 to 2
I0306 13:23:26.500384       6 event.go:217] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"prod", Name:"ganalytics-prod-3154642894", UID:"14aceba8-0270-11e7-9692-12b3f43365d0", APIVersion:"extensions", ResourceVersion:"9142085", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: ganalytics-prod-3154642894-rgxpp
I0306 13:23:26.503913       6 event.go:217] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"prod", Name:"ganalytics-prod-3154642894", UID:"14aceba8-0270-11e7-9692-12b3f43365d0", APIVersion:"extensions", ResourceVersion:"9142085", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: ganalytics-prod-3154642894-2bn4n
I0306 13:23:26.505992       6 deployment_controller.go:298] Error syncing deployment prod/ganalytics-prod: Operation cannot be fulfilled on deployments.extensions "ganalytics-prod": the object has been modified; please apply your changes to the latest version and try again
I0306 13:23:52.359972       6 event.go:217] Event(api.ObjectReference{Kind:"Deployment", Namespace:"prod", Name:"ganalytics-prod", UID:"524047a6-001c-11e7-bef2-0ab9b20c0a56", APIVersion:"extensions", ResourceVersion:"9142109", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set ganalytics-prod-2180515786 to 1
I0306 13:23:52.360664       6 replica_set.go:512] Too many "prod"/"ganalytics-prod-2180515786" replicas, need 1, deleting 1
I0306 13:23:52.360726       6 controller_utils.go:523] Controller ganalytics-prod-2180515786 deleting pod prod/ganalytics-prod-2180515786-lbwjj
I0306 13:23:52.412882       6 deployment_controller.go:298] Error syncing deployment prod/ganalytics-prod: Operation cannot be fulfilled on deployments.extensions "ganalytics-prod": the object has been modified; please apply your changes to the latest version and try again
I0306 13:23:52.413032       6 event.go:217] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"prod", Name:"ganalytics-prod-2180515786", UID:"6412efe7-0249-11e7-9692-12b3f43365d0", APIVersion:"extensions", ResourceVersion:"9142165", FieldPath:""}): type: 'Normal' reason: 'SuccessfulDelete' Deleted pod: ganalytics-prod-2180515786-lbwjj
I0306 13:23:52.456975       6 deployment_controller.go:298] Error syncing deployment prod/ganalytics-prod: Operation cannot be fulfilled on deployments.extensions "ganalytics-prod": the object has been modified; please apply your changes to the latest version and try again
I0306 13:23:52.474537       6 deployment_controller.go:298] Error syncing deployment prod/ganalytics-prod: Operation cannot be fulfilled on deployments.extensions "ganalytics-prod": the object has been modified; please apply your changes to the latest version and try again
W0306 13:24:27.496025       6 reflector.go:319] pkg/controller/garbagecollector/garbagecollector.go:768: watch of <nil> ended with: 401: The event in requested index is outdated and cleared (the requested history has been cleared [9141247/9141109]) [9142246]
```
**What you expected to happen**:
The old ReplicaSet should have scaled down to 0

**How to reproduce it** (as minimally and precisely as possible):
I only seem to have it with this particular deploy... Only changes to trigger this deployment was an image change.

**Anything else we need to know**:
All above examples are for the deployment ganalytics-prod, I have a second deployment managed by the same chart named ganalytics-prod-worker that uses the same Docker image but different configurations, which has all the same symptoms. Both have done this twice today.

**Comment by ravilr** [comment]
possibly dupe of https://github.com/kubernetes/kubernetes/issues/41641 
@kargakis made some fixes(linked from above issue) which are due to be released in v1.5.4.

**Comment by kargakis** [comment]
Can you post the output of `kubectl get deploy,rs -o yaml` next time it occurs? Highly likely that it's 41641 but I want to make sure.

**Comment by bcorijn** [comment]
Ah, my search-fu let me down, I had not bumped into 41641 but we indeed use a rather high MinReadySeconds on this deployment. 
Here's the (slightly redacted) output after triggering another deploy and seeing this issue again:
```
~ $ kubectl get deploy,rs -o yaml --namespace=prod -l app=ganalytics,role=main
apiVersion: v1
items:
- apiVersion: extensions/v1beta1
  kind: Deployment
  metadata:
    annotations:
      deployment.kubernetes.io/revision: "4"
    creationTimestamp: 2017-03-03T14:18:49Z
    generation: 4
    labels:
      app: ganalytics
      chart: ganalytics-1
      heritage: Tiller
      release: ganalytics-prod
      role: main
    name: ganalytics-prod
    namespace: prod
    resourceVersion: "9272337"
    selfLink: /apis/extensions/v1beta1/namespaces/prod/deployments/ganalytics-prod
    uid: 524047a6-001c-11e7-bef2-0ab9b20c0a56
  spec:
    minReadySeconds: 25
    replicas: 2
    revisionHistoryLimit: 5
    selector:
      matchLabels:
        app: ganalytics
        release: ganalytics-prod
        role: main
    strategy:
      rollingUpdate:
        maxSurge: 100%
        maxUnavailable: 0
      type: RollingUpdate
    template:
      metadata:
        creationTimestamp: null
        labels:
          app: ganalytics
          chart: ganalytics-1
          heritage: Tiller
          release: ganalytics-prod
          role: main
          version: v56
      spec:
        containers:
          env:
          - name: DUMMY
            value: ABCD
          image: example/ganalytics:v56
          imagePullPolicy: IfNotPresent
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: 3000
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 5
          name: ganalytics
          ports:
          - containerPort: 3000
            protocol: TCP
          resources:
            requests:
              cpu: 200m
              memory: 256M
          terminationMessagePath: /dev/termination-log
        dnsPolicy: ClusterFirst
        restartPolicy: Always
        securityContext: {}
        terminationGracePeriodSeconds: 30
  status:
    availableReplicas: 2
    conditions:
    - lastTransitionTime: 2017-03-07T07:01:23Z
      lastUpdateTime: 2017-03-07T07:01:23Z
      message: Deployment has minimum availability.
      reason: MinimumReplicasAvailable
      status: "True"
      type: Available
    observedGeneration: 4
    replicas: 3
    unavailableReplicas: 1
    updatedReplicas: 2
- apiVersion: extensions/v1beta1
  kind: ReplicaSet
  metadata:
    annotations:
      deployment.kubernetes.io/desired-replicas: "2"
      deployment.kubernetes.io/max-replicas: "4"
      deployment.kubernetes.io/revision: "3"
    creationTimestamp: 2017-03-06T13:23:26Z
    generation: 2
    labels:
      app: ganalytics
      chart: ganalytics-1
      heritage: Tiller
      pod-template-hash: "3154642894"
      release: ganalytics-prod
      role: main
      version: v56
    name: ganalytics-prod-3154642894
    namespace: prod
    resourceVersion: "9272336"
    selfLink: /apis/extensions/v1beta1/namespaces/prod/replicasets/ganalytics-prod-3154642894
    uid: 14aceba8-0270-11e7-9692-12b3f43365d0
  spec:
    minReadySeconds: 25
    replicas: 1
    selector:
      matchLabels:
        app: ganalytics
        pod-template-hash: "3154642894"
        release: ganalytics-prod
        role: main
    template:
      metadata:
        creationTimestamp: null
        labels:
          app: ganalytics
          chart: ganalytics-1
          heritage: Tiller
          pod-template-hash: "3154642894"
          release: ganalytics-prod
          role: main
          version: v56
      spec:
        containers:
          env:
          image: example/ganalytics:v56
          imagePullPolicy: IfNotPresent
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: 3000
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 5
          name: ganalytics
          ports:
          - containerPort: 3000
            protocol: TCP
          resources:
            requests:
              cpu: 200m
              memory: 256M
          terminationMessagePath: /dev/termination-log
        dnsPolicy: ClusterFirst
        restartPolicy: Always
        securityContext: {}
        terminationGracePeriodSeconds: 30
  status:
    availableReplicas: 1
    fullyLabeledReplicas: 1
    observedGeneration: 2
    readyReplicas: 1
    replicas: 1
- apiVersion: extensions/v1beta1
  kind: ReplicaSet
  metadata:
    annotations:
      deployment.kubernetes.io/desired-replicas: "2"
      deployment.kubernetes.io/max-replicas: "4"
      deployment.kubernetes.io/revision: "4"
    creationTimestamp: 2017-03-07T10:42:58Z
    generation: 1
    labels:
      app: ganalytics
      chart: ganalytics-1
      heritage: Tiller
      pod-template-hash: "317661612"
      release: ganalytics-prod
      role: main
      version: v56
    name: ganalytics-prod-317661612
    namespace: prod
    resourceVersion: "9272329"
    selfLink: /apis/extensions/v1beta1/namespaces/prod/replicasets/ganalytics-prod-317661612
    uid: d425221a-0322-11e7-9692-12b3f43365d0
  spec:
    minReadySeconds: 25
    replicas: 2
    selector:
      matchLabels:
        app: ganalytics
        pod-template-hash: "317661612"
        release: ganalytics-prod
        role: main
    template:
      metadata:
        creationTimestamp: null
        labels:
          app: ganalytics
          chart: ganalytics-1
          heritage: Tiller
          pod-template-hash: "317661612"
          release: ganalytics-prod
          role: main
          version: v56
      spec:
        containers:
          env:
          - name: DUMMY
            value: ABCD
          image: example/ganalytics:v56
          imagePullPolicy: IfNotPresent
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: 3000
              scheme: HTTP
            initialDelaySeconds: 30
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 5
          name: ganalytics
          ports:
          - containerPort: 3000
            protocol: TCP
          resources:
            requests:
              cpu: 200m
              memory: 256M
          terminationMessagePath: /dev/termination-log
        dnsPolicy: ClusterFirst
        restartPolicy: Always
        securityContext: {}
        terminationGracePeriodSeconds: 30
  status:
    availableReplicas: 1
    fullyLabeledReplicas: 2
    observedGeneration: 1
    readyReplicas: 2
    replicas: 2
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
```

**Comment by kargakis** [comment]
MinReadySeconds should be totally fine to use, this indeed is a dupe of #41641. Should be fixed in 1.5.4/1.6.0 but it will get even more robust in time.
